### ISSUE TITLE
Duplicate keys cause earlier TypedDict values to be skipped during type checking

### ISSUE THREAD

**Issue by Hmanbo** [python/mypy#21841]
**Bug Report**

When a dictionary literal used to construct a `TypedDict` contains duplicate
keys, mypy only type-checks the final value for that key. Earlier value
expressions are skipped entirely.

This can hide real errors because Python still evaluates every value expression
in the dictionary literal, including values that are later overwritten by a
duplicate key.

**To Reproduce**

```python
from typing import TypedDict


class Payload(TypedDict):
    value: int


def parse(value: str) -> int:
    return len(value)


payload = Payload({"value": parse(123), "value": 0})
```

```console
$ mypy --show-error-codes --no-incremental example.py
Success: no issues found in 1 source file
```

Running the program demonstrates that the skipped expression is evaluated:

```console
TypeError: object of type 'int' has no len()
```

**Expected Behavior**

Mypy should type-check every value expression in the dictionary literal, even
if a later duplicate key determines the final value of the `TypedDict` field.
In this example, it should report an error similar to:

```text
error: Argument 1 to "parse" has incompatible type "int"; expected "str"  [arg-type]
```

The same expression is correctly rejected when the target is a regular
`dict[str, int]`. The issue also occurs with a contextually typed literal:

```python
payload: Payload = {"value": parse(123), "value": 0}
```

**Actual Behavior**

Mypy reports no issues because only the last value associated with the
duplicate key is checked.

**Additional Context**

The current implementation replaces the previously collected expression
without type-checking it and has a TODO to type-check all duplicate-key values:

https://github.com/python/mypy/blob/6eb141f03a8b43c274f15cc4978d617a71d8f37e/mypy/checkexpr.py#L839-L842

PR #15425 also mentioned repeated `TypedDict` keys as a known limitation, but I
could not find an existing issue that tracks it:

https://github.com/python/mypy/pull/15425

**Environment Used to Reproduce**

- Mypy versions used:
  - `mypy 2.3.0 (compiled: yes)`
  - `mypy 2.4.0+dev.6eb141f03a8b43c274f15cc4978d617a71d8f37e (compiled: no)`
- Mypy command-line flags: `--show-error-codes --no-incremental`
- Mypy configuration options: none
- Python version used: 3.12.8

**Comment by Hmanbo** [comment]
I'd like to work on this.

My plan is to preserve all explicit values associated with a duplicate key
for type checking, while keeping the final value first for TypedDict generic
type inference. I'll also add regression tests for both TypedDict constructor
calls and contextually typed dictionary literals.

Does this approach sound reasonable?
